### PR TITLE
swarm/network: simplified neighbourhood depth calc

### PR DESCRIPTION
**Pull request by acud**
this PR simplifies the neighbourhood depth calculation for the kademlia and reuses the same logic in the peer pot map creation

### CODE REVIEW

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -451,11 +451,14 @@ func (k *Kademlia) neighbourhoodDepth() (depth int) {
 	var lastPo int
 
 	f := func(v pot.Val, i int) bool {
+		if bytes.Equal(pot.ToBytes(v), pivotAddr) {
```
Comment (nolash) [inline swarm/network/kademlia.go]: Under what condition should we be getting ourselves in these iterations. Isn't that actually an anomaly?

File: swarm/network/kademlia.go [hunk 2]
```diff
@@ -492,6 +495,10 @@ func (k *Kademlia) neighbourhoodDepth() (depth int) {
 	return depth
 }
 
+func (k *Kademlia) neighbourhoodDepth() (depth int) {
```
Comment (nolash) [inline swarm/network/kademlia.go]: Why do we need this now? Can't we just call `depthForPot` directly by `kademlia.NeighbourhoodDepth`?

File: swarm/network/kademlia.go [hunk 3]
```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
...
-		depthTraversed := false
-
-		np.EachNeighbour(addrs[i], pof, func(val pot.Val, po int) bool {
-			a := val.([]byte)
-
-			// 256 is self. We are selfless
-			if po == 256 {
+		np.EachNeighbour(a, pof, func(val pot.Val, po int) bool {
+			addr := val.([]byte)
+			if bytes.Equal(pot.ToBytes(val), a) {
```
Comment (nolash) [inline swarm/network/kademlia.go]: Same as above line 454

File: swarm/network/kademlia.go [hunk 4]
```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
 	ppmap := make(map[string]*PeerPot)
 
 	for i, a := range addrs {
-
 		// set to proxbin depth when all nn-peers are found
```
Comment (nolash) [inline swarm/network/kademlia.go]: This comment seems wrong now

File: swarm/network/kademlia.go [hunk 5]
```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
...
-		pl := 256
-
+		depth := depthForPot(np, kadMinProxSize, a)
 		// upon entering a new iteration
 		// this will hold the value the po should be
 		// if it's one higher than the po in the last iteration
-		prev := 256
+		prevPo := 256
 
 		// all bins outside proxbin depth with no peers
```
Comment (nolash) [inline swarm/network/kademlia.go]: This comment also seems imprecise now.

File: swarm/network/kademlia.go [hunk 6]
```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
...
-
-			// if true then all nn-bins have been filled
-			// start counting pl and set prev to current po initially (which will skip next block)
-			if pl == 256 && len(nns) >= kadMinProxSize {
-				pl = po
-				prev = po
+			// iterate through the neighbours, going from the closest to the farthest
+			// we calculate the nearest neighbours that should be in the set
+			// depth in all bins that are higher or equal than depth there are
+			// at least minproxbin size connected and depth-1 is not empty
```
Comment (nolash) [inline swarm/network/kademlia.go]: The last two lines are hard to understand. Please review and clarify.

File: swarm/network/kademlia.go [hunk 7]
```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
...
+			if po >= depth {
+				nns = append(nns, addr)
+				prevPo = depth - 1
```
Comment (nolash) [inline swarm/network/kademlia.go:663]: Why is this needed in addition to line 668? Can the logic be merged?

**Review by nolash — CHANGES_REQUESTED** [review]

File: swarm/network/kademlia.go (hunk 1 above)
Comment (acud) [reply swarm/network/kademlia.go]: @nolash this is exactly the same as was removed on the left side L587:
```
			if po == 256 {	
```
...
@zelig am I correct?

File: swarm/network/kademlia.go (hunk 7 above)
Comment (acud) [reply swarm/network/kademlia.go:663]: you have to return here otherwise you'd get empty bins inside the neighbourhood depth. i don't see how the two could be merged. seems about right.

File: swarm/network/kademlia.go (hunk 1 above)
Comment (zelig) [reply swarm/network/kademlia.go]: maybe going back to `po == 256` is simpler

**Review by zelig — APPROVED** [review]

File: swarm/network/kademlia.go (hunk 1 above)
Comment (nolash) [reply swarm/network/kademlia.go]: @justelad the question is about the iterator, not the value itself. Should the iterator _ever_ return ourselves?

Comment (acud) [reply swarm/network/kademlia.go]: good question. i guess _not_ since the method name is `EachNeighbour` and `self` is not a neighbour. i'll make the change. good point 👍

Comment (acud) [reply swarm/network/kademlia.go]: @nolash, come to think of it, that method name could be actually changed to `EachPeer`, since it refers to actual peers connected, not just neighbours.
@zelig, I noted to @nolash that some of the terminology in the kademlia is somewhat misleading (as I already told you about `minProxBinSize`), `EachNeighbour` is the same case - we iterate over peers not neighbours, the iterator in this case should not have a notion of what is neighbouring and what is not, and that is up to the iteratee to assert. (my 2 cents)

**Comment by nolash** [comment]
@justelad Tests fail; `neighbourhoodDepth` calls not replaced with `depthForPot` it seems...

I pushed a minor amend in comments, hope it's ok.

**Review by nolash — APPROVED** [review]
Approved provided tests pass
